src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs

```diff
@@ -66,68 +66,181 @@ public static int IndexOf(ref char searchSpace, int searchSpaceLength, ref char
-            IntPtr i = (IntPtr)0; // Use IntPtr for arithmetic to avoid unnecessary 64->32->64 truncations
+            int minLength = (firstLength < secondLength) ? firstLength : secondLength;
+            int offset = 0;
+
+            if (Sse2.IsSupported && minLength >= Vector128<ushort>.Count)
+            {
+                // >= Sse2 intrinsics are supported, and length is enough to use them so use that path.
+                // We jump forward to the intrinsics at the end of them method so a naive branch predict
+                // will choose the non-intrinsic path so short lengths which don't gain anything aren't
+                // overly disadvantaged by having to jump over a lot of code. Whereas the longer lengths
+                // more than make this back from the intrinsics.
+                goto IntrinsicsCompare;
+            }
 
-            if ((byte*)minLength >= (byte*)(sizeof(UIntPtr) / sizeof(char)))
+            // Only use the Vector path if intrinsics aren't supported.
+            if (!Sse2.IsSupported && Vector.IsHardwareAccelerated)
             {
-                if (Vector.IsHardwareAccelerated && (byte*)minLength >= (byte*)Vector<ushort>.Count)
+                int lengthToExamine = minLength - Vector<ushort>.Count;
+                if (lengthToExamine >= 0)
                 {
-                    IntPtr nLength = minLength - Vector<ushort>.Count;
                     do
                     {
-                        if (Unsafe.ReadUnaligned<Vector<ushort>>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref first, i))) !=
-                            Unsafe.ReadUnaligned<Vector<ushort>>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref second, i))))
+                        if (LoadVector(ref firstStart, offset) != LoadVector(ref secondStart, offset))
                         {
-                            break;
+                            goto CharwiseCompare;
                         }
-                        i += Vector<ushort>.Count;
-                    }
-                    while ((byte*)nLength >= (byte*)i);
-                }
+                        offset += Vector<ushort>.Count;
+                    } while (lengthToExamine > offset);
 
-                while ((byte*)minLength >= (byte*)(i + sizeof(UIntPtr) / sizeof(char)))
-                {
-                    if (Unsafe.ReadUnaligned<UIntPtr>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref first, i))) !=
-                        Unsafe.ReadUnaligned<UIntPtr>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref second, i))))
-                    {
-                        break;
-                    }
-                    i += sizeof(UIntPtr) / sizeof(char);
+                    goto CharwiseCompare;
                 }
             }
 
+        CharwiseCompare:
 #if BIT64
-            if ((byte*)minLength >= (byte*)(i + sizeof(int) / sizeof(char)))
+            if ((byte*)minLength >= (byte*)(offset + sizeof(int) / sizeof(char)))
```

Comment:
Why are these compared as pointers?

Reply:
Hmm; I think asm regression has crept in across PR versions as this PR's version will have 64->32->64 movs.

Cleaning up